### ISSUE TITLE
Ghost V4 to V4.1: much slower post rendering time (x10 increase)

### ISSUE THREAD

**Issue by valeriansaliou** [TryGhost/Ghost#12791]
### Issue Summary

Hello there! First of all thanks for Ghost 4, looks super nice so far. Great work.

I'm reporting a concern I have after upgrading from a self-hosted version of Ghost 3 (latest) to Ghost 4, where page rendering performance has severely degraded.

While, w/ V3, pages rendered in less than 200ms and did not show much CPU activity on `htop`, now that I am running Ghost V4, pages take much longer to render (2-3 seconds), and I can clearly see 1 CPU core at 100% for the same time it takes to render the page, as reported by `htop`. That 100% comes from the NodeJS process running Ghost, and not MySQL nor blocking disk I/O.

Note that pages render fine, and nothing looks broken. They just take x10 the time they took to render in Ghost V3.

Also, this slowdown only seem to affect pages that contain posts as well as pages listing posts. Simple pages (non-posts) seem to be as fast as before (millisecond range).

I'm using a theme that is rather lightweight and standard (not doing much templating work), and migrated it over to Ghost V4 full compatibility (no warnings nor errors are being reported by GScan).

Note that I am running Ghost on an ARM CPU, which is obviously not as fast as server Intel Xeon CPUs, but a x10 performance degradation makes me wonder something could be optimized, and even show gains for Ghost V4 instances running on traditional Intel CPUs.

As poor performance impacts SEO, I'm concerned that this negatively impacts the position in Google of a lot of Ghost blogs that just migrated to V4.

### To Reproduce

- Measure page rendering performance in Ghost V3 (in my case: ~200ms on an ARM core)
- Upgrade from Ghost V3 to Ghost V4
- Measure page rendering performance in Ghost V4 (takes x10 time on the same ARM core)

### Technical details:

* Ghost Version: `v4.1.0`
* Node Version: `v10.23.0`
* Browser/OS: Firefox on macOS
* Database: `mysql  Ver 15.1 Distrib 10.3.27-MariaDB, for debian-linux-gnueabihf (armv8l) using readline 5.2`

**Comment by joseftw** [comment]
Seeing the same thing, im also selfhosted.
Ghost Version: 4.1.0

**Comment by iBotPeaches** [comment]
I'm going to add my experience here as well. I jumped to v4 last night, before on latest v3 - average around 100-300ms between index page.

Post upgrade, its now 900ms-3.1s. This is the very first request of the blog. My theme outside of the upgrades needed to pass gscan is the same. 

For me, I did take this chance to v4 to also jump from node `v10` to `v14`, but I even reverted that back to check and still remains slower than before.

So I went off to get some profiling information.
```
➜  Node ab -n100 -c 10 https://connortumbleson.com/
Server Software:        nginx
Server Hostname:        connortumbleson.com
Server Port:            443
SSL/TLS Protocol:       TLSv1.2,ECDHE-RSA-AES256-GCM-SHA384,2048,256

Document Path:          /
Document Length:        35177 bytes

Concurrency Level:      10
Time taken for tests:   96.214 seconds
Complete requests:      100
Failed requests:        0
Total transferred:      3561200 bytes
HTML transferred:       3517700 bytes
Requests per second:    1.04 [#/sec] (mean)
Time per request:       9621.410 [ms] (mean)
Time per request:       962.141 [ms] (mean, across all concurrent requests)
Transfer rate:          36.15 [Kbytes/sec] received
```

Not too good. So I went off to start a profiler and do the same test again.
```
NODE_ENV=production node --prof /usr/bin/ghost run
```

```
Statistical profiling result from isolate-0x643f030-6304-v8.log, (1435 ticks, 1 unaccounted, 0 excluded).

 [Shared libraries]:
   ticks  total  nonlib   name
   1221   85.1%          /usr/bin/node
    110    7.7%          /lib/x86_64-linux-gnu/libc-2.31.so
     61    4.3%          /lib/x86_64-linux-gnu/libpthread-2.31.so
     14    1.0%          /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.28

 [JavaScript]:
   ticks  total  nonlib   name
     15    1.0%   51.7%  LazyCompile: *resolve path.js:973:10
      5    0.3%   17.2%  LazyCompile: *Module._nodeModulePaths internal/modules/cjs/loader.js:575:37
      3    0.2%   10.3%  LazyCompile: *normalizeString path.js:52:25
      2    0.1%    6.9%  LazyCompile: *realpathSync fs.js:1650:22
      1    0.1%    3.4%  RegExp: (?:^|\/)\.?\.$
      1    0.1%    3.4%  LazyCompile: *hidden internal/errors.js:286:25
      1    0.1%    3.4%  LazyCompile: *<anonymous> internal/per_context/primordials.js:23:10

 [C++]:
   ticks  total  nonlib   name

 [Summary]:
   ticks  total  nonlib   name
     28    2.0%   96.6%  JavaScript
      0    0.0%    0.0%  C++
     55    3.8%  189.7%  GC
   1406   98.0%          Shared libraries
      1    0.1%          Unaccounted

 [C++ entry points]:
   ticks    cpp   total   name

 [Bottom up (heavy) profile]:
  Note: percentage shows a share of a particular caller in the total
  amount of its parent calls.
  Callers occupying less than 1.0% are not shown.

   ticks parent  name
   1221   85.1%  /usr/bin/node
    938   76.8%    /usr/bin/node
    384   40.9%      LazyCompile: ~wrapSafe internal/modules/cjs/loader.js:964:18
    384  100.0%        LazyCompile: ~Module._compile internal/modules/cjs/loader.js:1017:37
    380   99.0%          LazyCompile: ~Module._extensions..js internal/modules/cjs/loader.js:1072:37
    380  100.0%            LazyCompile: ~Module.load internal/modules/cjs/loader.js:916:33
      4    1.0%          Eval: ~<anonymous> /var/www/connor/blog/versions/4.1.0/node_modules/jsonpath/lib/aesprim.js:1:1
      4  100.0%            LazyCompile: ~Module._compile internal/modules/cjs/loader.js:1017:37
     88    9.4%      LazyCompile: ~compileForInternalLoader internal/bootstrap/loaders.js:270:27
     75   85.2%        LazyCompile: ~nativeModuleRequire internal/bootstrap/loaders.js:303:29
 ```

---
 
 I can't pretend to understand that, so I took a backup and theme and went to v3 latest to run this test again. I did do this locally though so I wouldn't affect my production setup, so I'm willing to toss this entire test if you think it isn't valid.
 
 ```
 ➜  ~ ab -n100 -c 10 http://localhost:2368/
This is ApacheBench, Version 2.3 <$Revision: 1706008 $>
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking localhost (be patient).....done


Server Software:        
Server Hostname:        localhost
Server Port:            2368

Document Path:          /
Document Length:        33098 bytes

Concurrency Level:      10
Time taken for tests:   18.994 seconds
Complete requests:      100
Failed requests:        2
   (Connect: 0, Receive: 0, Length: 2, Exceptions: 0)
Total transferred:      3335326 bytes
HTML transferred:       3309126 bytes
Requests per second:    5.26 [#/sec] (mean)
Time per request:       1899.405 [ms] (mean)
Time per request:       189.940 [ms] (mean, across all concurrent requests)
Transfer rate:          171.48 [Kbytes/sec] received
```
5.26r/sec vs 1.04/sec so that is better. What about the profile test?

```
Statistical profiling result from isolate-0x439fe10-29061-v8.log, (24131 ticks, 193 unaccounted, 0 excluded).

 [Shared libraries]:
   ticks  total  nonlib   name
  14041   58.2%          /usr/bin/node
   1480    6.1%          /lib/x86_64-linux-gnu/libpthread-2.23.so
    591    2.4%          /lib/x86_64-linux-gnu/libc-2.23.so
     20    0.1%          /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.28
     12    0.0%          [vdso]
      8    0.0%          /lib/x86_64-linux-gnu/libm-2.23.so

 [JavaScript]:
   ticks  total  nonlib   name
    113    0.5%    1.4%  LazyCompile: *resolve path.js:973:10
     84    0.3%    1.1%  LazyCompile: *parse /home/ibotpeaches/Projects/Node/ghost3/versions/3.42.3/node_modules/handlebars/dist/cjs/handlebars/compiler/parser.js:269:30
     78    0.3%    1.0%  LazyCompile: *baseClone /home/ibotpeaches/Projects/Node/ghost3/versions/3.42.3/node_modules/lodash/lodash.js:2620:23
     77    0.3%    1.0%  LazyCompile: *normalizeString path.js:52:25
     64    0.3%    0.8%  LazyCompile: *next /home/ibotpeaches/Projects/Node/ghost3/versions/3.42.3/node_modules/handlebars/dist/cjs/handlebars/compiler/parser.js:478:32
     45    0.2%    0.6%  LazyCompile: *parse /home/ibotpeaches/Projects/Node/ghost3/versions/3.42.3/node_modules/jsonpath/generated/parser.js:169:22
     38    0.2%    0.5%  LazyCompile: *wrap /home/ibotpeaches/Projects/Node/ghost3/versions/3.42.3/node_modules/handlebars/dist/cjs/handlebars/compiler/code-gen.js:101:22
     37    0.2%    0.5%  LazyCompile: *Tokenizer._parse /home/ibotpeaches/Projects/Node/ghost3/versions/3.42.3/node_modules/htmlparser2/lib/Tokenizer.js:702:38
     32    0.1%    0.4%  LazyCompile: *readPackage internal/modules/cjs/loader.js:231:21
     29    0.1%    0.4%  RegExp: <%-([\s\S]+?)%>|<%=([\s\S]+?)%>|\$\{([^\\}]*(?:\\.[^\\}]*)*)\}|<%([\s\S]+?)%>|$
     29    0.1%    0.4%  LazyCompile: *castChunk /home/ibotpeaches/Projects/Node/ghost3/versions/3.42.3/node_modules/handlebars/dist/cjs/handlebars/compiler/code-gen.js:52:19
     28    0.1%    0.4%  LazyCompile: *ExpressHbs.compile /home/ibotpeaches/Projects/Node/ghost3/versions/3.42.3/node_modules/express-hbs/lib/hbs.js:329:40
     24    0.1%    0.3%  LazyCompile: *moment.updateOffset /home/ibotpeaches/Projects/Node/ghost3/versions/3.42.3/node_modules/moment-timezone/moment-timezone.js:533:33
     24    0.1%    0.3%  LazyCompile: *configFromStringAndFormat /home/ibotpeaches/Projects/Node/ghost3/versions/3.42.3/node_modules/moment/moment.js:2405:39
     24    0.1%    0.3%  LazyCompile: *anonymous /home/ibotpeaches/Projects/Node/ghost3/versions/3.42.3/node_modules/handlebars/dist/cjs/handlebars/compiler/parser.js:12:42
     24    0.1%    0.3%  LazyCompile: *<anonymous> /home/ibotpeaches/Projects/Node/ghost3/versions/3.42.3/node_modules/lodash.defaults/index.js:229:27
     24    0.1%    0.3%  LazyCompile: *<anonymous> /home/ibotpeaches/Projects/Node/ghost3/versions/3.42.3/node_modules/express-hbs/lib/hbs.js:162:25
     20    0.1%    0.3%  LazyCompile: *module.exports /home/ibotpeaches/Projects/Node/ghost3/versions/3.42.3/node_modules/cheerio/lib/cheerio.js:30:40
     20    0.1%    0.3%  LazyCompile: *compile /home/ibotpeaches/Projects/Node/ghost3/versions/3.42.3/node_modules/handlebars/dist/cjs/handlebars/compiler/javascript-compiler.js:73:28
     20    0.1%    0.3%  LazyCompile: *baseClone /home/ibotpeaches/Projects/Node/ghost3/versions/3.42.3/node_modules/bookshelf/node_modules/lodash/lodash.js:2620:23
     19    0.1%    0.2%  LazyCompile: *generateList /home/ibotpeaches/Projects/Node/ghost3/versions/3.42.3/node_modules/handlebars/dist/cjs/handlebars/compiler/code-gen.js:143:38
     18    0.1%    0.2%  LazyCompile: *SourceNode_add /home/ibotpeaches/Projects/Node/ghost3/versions/3.42.3/node_modules/source-map/lib/source-node.js:172:51
     18    0.1%    0.2%  LazyCompile: *Module._nodeModulePaths internal/modules/cjs/loader.js:567:37
     18    0.1%    0.2%  LazyCompile: *<anonymous> /home/ibotpeaches/Projects/Node/ghost3/versions/3.42.3/node_modules/lodash.defaults/index.js:174:18
     17    0.1%    0.2%  LazyCompile: *parse url.js:158:37
     17    0.1%    0.2%  LazyCompile: *Module._load internal/modules/cjs/loader.js:648:24
     16    0.1%    0.2%  LazyCompile: *each /home/ibotpeaches/Projects/Node/ghost3/versions/3.42.3/core/server/models/base/index.js:479:36
     16    0.1%    0.2%  LazyCompile: *configFromArray /home/ibotpeaches/Projects/Node/ghost3/versions/3.42.3/node_modules/moment/moment.js:2050:30
     15    0.1%    0.2%  LazyCompile: *readFileSync fs.js:361:22
     15    0.1%    0.2%  LazyCompile: *normalize path.js:1004:12
     15    0.1%    0.2%  LazyCompile: *createDuration /home/ibotpeaches/Projects/Node/ghost3/versions/3.42.3/node_modules/moment/moment.js:3027:29
     15    0.1%    0.2%  LazyCompile: *basePick /home/ibotpeaches/Projects/Node/ghost3/versions/3.42.3/node_modules/lodash/lodash.js:3759:22
```

So then I thought. Lets upgrade my local test v3 to v4 to keep all the variables the same this time and do the test again. Since the blog needed to do migrations, etc. I upgraded it. Let it sit for 60s and then stopped and restarted for profiling.

So Apache Workbench test between local v3/v4.

```
➜  ~ ab -n100 -c 10 http://localhost:2368/
This is ApacheBench, Version 2.3 <$Revision: 1706008 $>
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking localhost (be patient).....done


Server Software:        
Server Hostname:        localhost
Server Port:            2368

Document Path:          /
Document Length:        34602 bytes

Concurrency Level:      10
Time taken for tests:   32.017 seconds
Complete requests:      100
Failed requests:        1
   (Connect: 0, Receive: 0, Length: 1, Exceptions: 0)
Total transferred:      3486063 bytes
HTML transferred:       3459863 bytes
Requests per second:    3.12 [#/sec] (mean)
Time per request:       3201.727 [ms] (mean)
Time per request:       320.173 [ms] (mean, across all concurrent requests)
Transfer rate:          106.33 [Kbytes/sec] received
```

So same everything now. v4 is `3.12r/second` while v3 is `5.26r/second`. So now lets look at the profile.

```
Statistical profiling result from isolate-0x4453e10-29891-v8.log, (33039 ticks, 306 unaccounted, 0 excluded).

 [Shared libraries]:
   ticks  total  nonlib   name
  22237   67.3%          /usr/bin/node
    978    3.0%          /lib/x86_64-linux-gnu/libc-2.23.so
    168    0.5%          /lib/x86_64-linux-gnu/libpthread-2.23.so
     42    0.1%          /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.28
     18    0.1%          [vdso]
      9    0.0%          /lib/x86_64-linux-gnu/libm-2.23.so

 [JavaScript]:
   ticks  total  nonlib   name
    166    0.5%    1.7%  LazyCompile: *Tokenizer._parse /home/ibotpeaches/Projects/Node/ghost3/versions/4.1.0/node_modules/htmlparser2/lib/Tokenizer.js:702:38
    117    0.4%    1.2%  LazyCompile: *resolve path.js:973:10
    116    0.4%    1.2%  LazyCompile: *module.exports /home/ibotpeaches/Projects/Node/ghost3/versions/4.1.0/node_modules/cheerio/lib/cheerio.js:30:40
    116    0.4%    1.2%  LazyCompile: *<anonymous> /home/ibotpeaches/Projects/Node/ghost3/versions/4.1.0/node_modules/lodash.defaults/index.js:229:27
    110    0.3%    1.1%  LazyCompile: *parse /home/ibotpeaches/Projects/Node/ghost3/versions/4.1.0/node_modules/parse-entities/index.js:115:15
     92    0.3%    1.0%  LazyCompile: *createWrap /home/ibotpeaches/Projects/Node/ghost3/versions/4.1.0/node_modules/lodash.bind/index.js:738:20
     91    0.3%    0.9%  LazyCompile: *paragraph /home/ibotpeaches/Projects/Node/ghost3/versions/4.1.0/node_modules/remark-parse/lib/tokenize/paragraph.js:17:19
     83    0.3%    0.9%  LazyCompile: *parse /home/ibotpeaches/Projects/Node/ghost3/versions/4.1.0/node_modules/handlebars/dist/cjs/handlebars/compiler/parser.js:269:30
     75    0.2%    0.8%  LazyCompile: *<anonymous> /home/ibotpeaches/Projects/Node/ghost3/versions/4.1.0/node_modules/lodash.defaults/index.js:174:18
     70    0.2%    0.7%  RegExp: ^\s*|\s*$
     70    0.2%    0.7%  LazyCompile: *text /home/ibotpeaches/Projects/Node/ghost3/versions/4.1.0/node_modules/remark-parse/lib/tokenize/text.js:5:14
     69    0.2%    0.7%  LazyCompile: *normalizeString path.js:52:25
     58    0.2%    0.6%  LazyCompile: *next /home/ibotpeaches/Projects/Node/ghost3/versions/4.1.0/node_modules/handlebars/dist/cjs/handlebars/compiler/parser.js:478:32
     45    0.1%    0.5%  LazyCompile: *wrap /home/ibotpeaches/Projects/Node/ghost3/versions/4.1.0/node_modules/handlebars/dist/cjs/handlebars/compiler/code-gen.js:101:22
     43    0.1%    0.4%  LazyCompile: *baseClone /home/ibotpeaches/Projects/Node/ghost3/versions/4.1.0/node_modules/lodash/lodash.js:2662:23
     40    0.1%    0.4%  LazyCompile: *parse /home/ibotpeaches/Projects/Node/ghost3/versions/4.1.0/node_modules/jsonpath/generated/parser.js:169:22
     36    0.1%    0.4%  LazyCompile: *castChunk /home/ibotpeaches/Projects/Node/ghost3/versions/4.1.0/node_modules/handlebars/dist/cjs/handlebars/compiler/code-gen.js:52:19
     32    0.1%    0.3%  LazyCompile: *SourceNode_walk /home/ibotpeaches/Projects/Node/ghost3/versions/4.1.0/node_modules/source-map/lib/source-node.js:221:53
     31    0.1%    0.3%  RegExp: ^\s*|\s*$ {1}
     31    0.1%    0.3%  LazyCompile: *tokenize /home/ibotpeaches/Projects/Node/ghost3/versions/4.1.0/node_modules/remark-parse/lib/tokenizer.js:10:20
     30    0.1%    0.3%  LazyCompile: *initialize /home/ibotpeaches/Projects/Node/ghost3/versions/4.1.0/node_modules/cheerio/lib/static.js:24:28
     29    0.1%    0.3%  LazyCompile: *parseSelector /home/ibotpeaches/Projects/Node/ghost3/versions/4.1.0/node_modules/css-what/index.js:91:23
     28    0.1%    0.3%  LazyCompile: *ExpressHbs.compile /home/ibotpeaches/Projects/Node/ghost3/versions/4.1.0/node_modules/express-hbs/lib/hbs.js:329:40
     27    0.1%    0.3%  LazyCompile: *onParseComplete internal/url.js:237:25
     27    0.1%    0.3%  LazyCompile: *now /home/ibotpeaches/Projects/Node/ghost3/versions/4.1.0/node_modules/remark-parse/lib/tokenizer.js:126:17
     27    0.1%    0.3%  LazyCompile: *normalize path.js:1004:12
```

I'm attaching both full profiles.
[processed-local-v4.txt](https://github.com/TryGhost/Ghost/files/6170744/processed-local-v4.txt)
[processed-local-v3.txt](https://github.com/TryGhost/Ghost/files/6170745/processed-local-v3.txt)

**Comment by kevinansfield** [comment]
Thanks for the report. I've verified this and the cause is the changes from https://github.com/TryGhost/Ghost/pull/12787. I'm seeing unexpected behaviour with bookshelf.js calling the `format()` method multiple times for various posts on each render, the code paths in `format()` are known to be heavy but they are only expected to run once on save hence unexpected behaviour.

**Comment by denvergeeks** [comment]
Thank you (everyone) - I can add my voice to the choir. Seeing this is a known issue in progress saves me time and trouble.

**Comment by ErisDS** [comment]
To be absolutely clear. 4.0 had better performance, this is a bug introduced in 4.1. We recognise this bug is a  priority and it will be fixed in the next release. 

We do not recommend downgrading in the meantime. Just hold on to your hats if you can please!

**Comment by kevinansfield** [comment]
The performance fix has been released in 4.1.1 and is available to update to.
